libnavigation-core/src/androidTest/java/com/mapbox/navigation/core/trip/service/NativeNavigatorTest.kt

```diff
@@ -0,0 +1,314 @@
+    private lateinit var refreshedRouteResponse: String
+    private lateinit var callbackInvocations: MutableList<CallbackInvocation>
+
+    override fun setupMockLocation(): Location {
+        return mockLocationUpdatesRule.generateLocationUpdate {
+            latitude = startLocation.latitude()
+            longitude = startLocation.longitude()
+        }
+    }
+
+    @Before
+    fun setUp() {
+        runOnMainSync {
+            callbackInvocations = Collections.synchronizedList(mutableListOf<CallbackInvocation>())
+            mapboxNavigation = MapboxNavigationProvider.create(
+                NavigationOptions.Builder(activity)
+                    .accessToken(activity.getString(R.string.mapbox_access_token))
+                    .build()
+            )
+            // starts raw location updates - otherwise we don't get onStatus calls
+            mapboxNavigation.startTripSession()
+            val nativeNavigatorField = mapboxNavigation.javaClass.getDeclaredField("navigator")
+            nativeNavigatorField.isAccessible = true
+            val nativeNavigatorImpl =
+                nativeNavigatorField.get(mapboxNavigation) as MapboxNativeNavigatorImpl
+            nativeNavigatorField.isAccessible = false
+            val navigatorField = nativeNavigatorImpl.javaClass.getDeclaredField("navigator")
+            navigatorField.isAccessible = true
+            navigator = navigatorField.get(nativeNavigatorImpl) as Navigator
+            navigatorField.isAccessible = false
+            routes = DirectionsResponse.fromJson(
+                activity.resources.openRawResource(R.raw.route_response_route_refresh)
+                    .readBytes().decodeToString(),
+                RouteOptions.builder()
+                    .profile(DirectionsCriteria.PROFILE_DRIVING_TRAFFIC)
+                    .coordinatesList(
+                        listOf(
+                            startLocation,
+                            Point.fromLngLat(-121.480256, 38.576795)
+                        )
+                    )
+                    .build()
+            ).routes().toNavigationRoutes(RouterOrigin.Custom())
+            route = routes.first()
+            refreshedRouteResponse = activity.resources
+                .openRawResource(R.raw.route_response_route_refresh_annotations)
+                .readBytes().decodeToString()
+            multilegRoute = DirectionsRoute.fromJson(
+                activity.resources.openRawResource(R.raw.multileg_route)
+                    .readBytes().decodeToString()
+            ).toNavigationRoute(RouterOrigin.Custom())
+        }
+    }
+
+    @After
+    fun tearDown() {
+        MapboxNavigationProvider.destroy()
+    }
+
+    @Test
+    fun setRoutes() = runBlocking(Dispatchers.Main.immediate) {
+        navigator.addObserver { origin, _ ->
+            callbackInvocations.add(CallbackInvocation.Status(origin))
+        }
+        callbackInvocations.waitUntilHasSize(2)
```

Comment:
Just to imitate real-world behaviour and make sure we receive RoutesSet after there have already been some updates. IRL it would be FD -> wait some time until we get routes -> AG. I'll try to make it more clear.